### PR TITLE
Fix `<complex>` on NVRTC

### LINKED ISSUE #101: `cuda::std::complex` doesn't work with NVRTC because some `<cmath>` functions are missing
```
******************** TEST 'libcu++ :: std/numerics/complex.number/complex/types.pass.cpp' FAILED ********************
Compiled With: '/data/tmp/libcudacxx/utils/nvidia/nvrtc/nvrtc.sh  nvcc -o /data/tmp/libcudacxx/build/libcxx/test/std/numerics/complex.number/complex/Output/types.pass.cpp.o -x cu /data/tmp/libcudacxx/.upstream-tests/test/std/numerics/complex.number/complex/types.pass.cpp -c -v -ftemplate-depth=270 -ccbin=g++ -std=c++11 -include /data/tmp/libcudacxx/test/support/nasty_macros.h -I/data/tmp/libcudacxx/include -I/data/tmp/libcudacxx/build/libcxx/include/c++build -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D__STDC_CONSTANT_MACROS -Xcompiler -fno-exceptions -Xcompiler -fno-rtti -D_LIBCUDACXX_NO_RTTI -I/data/tmp/libcudacxx/test/support -I/data/tmp/libcudacxx/include -I/data/tmp/libcudacxx/../cuda/tools/cooperative_groups -I/data/tmp/libcudacxx/../cuda/tools/libcudacxxext --extended-lambda -gencode=arch=compute_61,code=sm_61 -gencode=arch=compute_61,code=compute_61 -Xcudafe --display_error_number -Werror all-warnings -D_LIBCUDACXX_DISABLE_PRAGMA_GCC_SYSTEM_HEADER -c && /data/tmp/libcudacxx/utils/nvidia/nvrtc/nvrtc.sh  nvcc -o /data/tmp/libcudacxx/build/libcxx/test/std/numerics/complex.number/complex/Output/types.pass.cpp.exe /data/tmp/libcudacxx/build/libcxx/test/std/numerics/complex.number/complex/Output/types.pass.cpp.o -v -ftemplate-depth=270 -ccbin=g++ -L/data/tmp/libcudacxx/build/lib -Xcompiler "-Wl,-rpath,/data/tmp/libcudacxx/build/lib" -L/data/tmp/libcudacxx/build/lib -Xcompiler "-Wl,-rpath,/data/tmp/libcudacxx/build/lib" -lm -lgcc_s -lgcc -lpthread -lrt -lc -lgcc_s -lgcc'
Command: ['/data/tmp/libcudacxx/build/libcxx/test/std/numerics/complex.number/complex/Output/types.pass.cpp.exe']
Exit Code: 1
Standard Output:
--
CUDA devices found: 1
Device 0: "TITAN Xp", Selected, SM61, 12787122176 [bytes]
/data/tmp/libcudacxx/include/cuda/std/detail/libcxx/include/cmath(324): error: the global scope has no "fpclassify"
/data/tmp/libcudacxx/include/cuda/std/detail/libcxx/include/cmath(325): error: the global scope has no "isnormal"
/data/tmp/libcudacxx/include/cuda/std/detail/libcxx/include/cmath(326): error: the global scope has no "isgreater"
/data/tmp/libcudacxx/include/cuda/std/detail/libcxx/include/cmath(327): error: the global scope has no "isgreaterequ
...
  libcu++ :: std/numerics/complex.number/complex.transcendentals/tan.pass.cpp
  libcu++ :: std/numerics/complex.number/complex.transcendentals/tanh.pass.cpp
  libcu++ :: std/numerics/complex.number/complex.value.ops/abs.pass.cpp
  libcu++ :: std/numerics/complex.number/complex.value.ops/arg.pass.cpp
  libcu++ :: std/numerics/complex.number/complex.value.ops/conj.pass.cpp
  libcu++ :: std/numerics/complex.number/complex.value.ops/imag.pass.cpp
  libcu++ :: std/numerics/complex.number/complex.value.ops/norm.pass.cpp
  libcu++ :: std/numerics/complex.number/complex.value.ops/polar.pass.cpp
  libcu++ :: std/numerics/complex.number/complex.value.ops/proj.pass.cpp
  libcu++ :: std/numerics/complex.number/complex.value.ops/real.pass.cpp
  libcu++ :: std/numerics/complex.number/complex/types.fail.cpp
  libcu++ :: std/numerics/complex.number/complex/types.pass.cpp
  libcu++ :: std/numerics/complex.number/layout.pass.cpp
  libcu++ :: std/utilities/utilities.general/broken_constant_eval.pass.cpp
Testing Time: 295.06s
  Unsupported      : 434
  Passed           : 421
  Expectedly Failed:   2
  Failed           :  76
# WALLTIME libcu++: 301.531 [sec]
```

### PR DESCRIPTION
**Pull request by wmaxey**
Changes made in this patchset to fix #101:

- `<cmath>` exposes a subset of `cuda::std::` math functions under NVRTC.
- `<cmath>` creates defines for `INFINITY` and `NAN` when NVRTC is detected
- Tests were reworked to no longer rely on a global test case struct. 
- A couple .fail tests required NVRTC exceptions.

### CODE REVIEW

File: .upstream-tests/test/std/numerics/complex.number/cases.h [hunk 1]
```diff
@@ -16,166 +16,200 @@
 #include <cuda/std/complex>
 #include <cuda/std/cassert>
 
-#ifdef __CUDA_ARCH__
```
Comment (wmaxey) [inline .upstream-tests/test/std/numerics/complex.number/cases.h:19]: This needs to be brought to attention, __CUDA_ARCH__ globals are not going to be supportable with NVC++. So this change serves a dual purpose in the future. 

Immediately, however, NVRTC cannot initialize the `testcases` constant using the dynamic floating point values from the likes of `CUDART_INF_F`.

File: libcxx/include/cmath [hunk 2]
```diff
@@ -312,33 +312,19 @@ long double    truncl(long double x);
 #pragma GCC system_header
 #endif
 
+#ifdef _LIBCUDACXX_COMPILER_NVRTC
+#include <math_constants.h>
```
Comment (leofang) [inline libcxx/include/cmath:316]: Just FYI: I am not sure what's the release strategy of libcu++, so maybe this can be ignored, but including this header has caused trouble in CuPy when a user runs CUDA + NVRTC + CuPy in a "runtime" environment (such as the CUDA Runtime Docker, or RAPIDS's favorite Conda env), because none of the standard CUDA headers are present there due to EULA.

Comment (wmaxey) [reply libcxx/include/cmath:316]: Thanks Leo!

That's insightful, I'm recalling now similar issues with cooperative_groups being unavailable in NVRTC. 

I have a more preferable solution now that I think about it.

File: .upstream-tests/test/std/numerics/complex.number/cases.h [hunk 3]
```diff
@@ -16,166 +16,200 @@
...
-    cuda::std::complex<double>(-0.5, INFINITY),
-    cuda::std::complex<double>(-0., INFINITY),
-    cuda::std::complex<double>(+0., INFINITY),
-    cuda::std::complex<double>(0.5, INFINITY),
-    cuda::std::complex<double>(1, INFINITY),
-    cuda::std::complex<double>(2, INFINITY),
-    cuda::std::complex<double>(INFINITY, INFINITY)
+using testcases_t = cuda::std::complex<double>[137];
+
+struct _testcases {
```
Comment (griwes) [inline .upstream-tests/test/std/numerics/complex.number/cases.h:21]: Does this exist to avoid needing the array overloads of being/end? If so, why is that needed? Do we just not expose them yet?

File: .upstream-tests/test/std/numerics/complex.number/cmplx.over/UDT_is_rejected.fail.cpp [hunk 4]
```diff
@@ -6,11 +6,14 @@
 //
 //===----------------------------------------------------------------------===//
 
+// UNSUPPORTED: nvrtc
```
Comment (griwes) [inline .upstream-tests/test/std/numerics/complex.number/cmplx.over/UDT_is_rejected.fail.cpp:9]: Here and in the other ones: add a comment explaining why it is unsupported.

File: .upstream-tests/test/std/numerics/complex.number/cmplx.over/UDT_is_rejected.fail.cpp [hunk 5]
```diff
@@ -6,11 +6,14 @@
...
+
 // <cuda/std/complex>
 
 // Test that UDT's convertible to an integral or floating point type do not
 // participate in overload resolution.
 
+
```
Comment (griwes) [inline .upstream-tests/test/std/numerics/complex.number/cmplx.over/UDT_is_rejected.fail.cpp:16]: Remove.

**Review by griwes — APPROVED** [review]
Approve pending resolution of the comments.

File: .upstream-tests/test/std/numerics/complex.number/cases.h (hunk 3 above)
Comment (wmaxey) [reply .upstream-tests/test/std/numerics/complex.number/cases.h:21]: I don't think we expose begin/end yet. Some tests might rely on range based for loops, so I just wanted satisfy that as non-obtrusively as possible.

NVRTC was particularly angry about dynamic initialization so I had to wrap up the array to make it work.

Comment (griwes) [reply .upstream-tests/test/std/numerics/complex.number/cases.h:21]: Cool, ship it!

File: libcxx/include/cmath (hunk 2 above)
Comment (leofang) [reply libcxx/include/cmath]: Wesley, the change in 6696f20 is cool 😄
